### ISSUE TITLE
Questions + issues with redux_store, and client vs server registration

### ISSUE THREAD

**Issue by vsai** [shakacode/react_on_rails#575]
Hi,
I'm new to ReactOnRails and still trying to figure out how some of the different pieces fit together.

1. Client vs server registration:
I have two files (similar to the tutorial): `client/app/bundles/test/startup/clientRegistration.jsx` and `client/app/bundles/test/startup/serverRegistration.jsx`.
Is serverRegistration only necessary when `prerender = true` for a component? Do we not need it otherwise?
Is clientRegistration always necessary? Or only if exists a component s.t. `prerender = false`?
When does this registration of components and stores actually happen? And some context for what they're actually doing?

2. Issues with redux_store + server vs client rendering
AFAIK server-side rendering only occurs if `prerender = true` for a component. Meaning HTML will be preloaded on the server.
My component has `prerender = false` (i.e. client-side rendering). Accordingly, I tried to initialize the redux_store in the client code:
 `const aStore = ReactOnRails.getStore('aStore');` and using `<Provider store= { aStore }><div><div/></Provider>`.
Without using a `redux_store` on the server. But I was getting an error:
`There are no stores hydrated and you are requesting the store aStore. This can happen if you are server rendering and you do not call redux_store near the top of your controller action's view (not the layout) and before any call to react_component.`

This was "fixed" by adding `redux_store("aStore", json_string)` to the controller action.
However, `aStore` was empty and wasn't initialized with the objects in `json_string`.
I have the Redux + React chrome extensions, and they both show that the store is empty.

Why is it necessary to "hydrate" the store on the server-side, using `redux_store`, if I'm not doing any server-side rendering for the components.
And even if I have to do that, shouldn't I be able to access any variables in the `json_string`?

**Comment by justin808** [comment]
1. Client vs Server registrations allows you to register different versions of the same component for the different bundles. If you are not pre-rendering (which you should not do at first), then you can ignore this. Follow the example from the tutorial first.
2. Please follow the examples. You need to decide if you want to register the store separately than the react component. There's no "server side" if you're not doing server rendering. You may have incorrectly setup the redux store in the view. You have to be careful with whether you are doing a `<%=` or `<%`. Again, please follow the examples closely.

If you'd like more help, we have coaching plans available where we'll screen share, review code, etc.: http://www.shakacode.com/work.

**Comment by vsai** [comment]
tl;dr: is there a better way to debug the issue I'm having? Is there a way to throw an error if `ReactOnRails.registerStore` doesn't actually register the store that is provided explaining why it wasn't able to register the store?

Thanks for the response. I took some time and went through the examples and through the code automatically created by the generators too.
Most of the code I have is similar to the code used in your react tutorial example. I have added it below for your reference, if you'd like to see it.

Is is there a way to better debug the issue I'm having? I know the clientRegistration is happening, and that the `App` is registered and appears in `ReactOnRails.registeredComponents()`. 

However, the `aStore` isn't being registered, and the `ReactOnRails.stores()` map is empty.
Before registering the `aStore`, I tried console.log the imported variable, and it correctly showed the contents of the imported "store".

`prerender = false`. `redux_store` isn't used anywhere. And there's no "server-side" code being run.

I'm not sure why ReactOnRails isn't registering the store, and even if it isn't, it'd be nice to be able to debug / throw an error when it fails to register the store that is passed in.


`app/controllers/as_controller.rb`
`
def index
    @as = As.all
end
`

`app/views/as/index.html.erb`:
`
<%= react_component("App", props: @selected, prerender: false, trace: true) %>
`

`startup/clientRegistration.jsx`:
`
import ReactOnRails from 'react-on-rails';

import App from './App';
import aStore from '../store/aStore';
ReactOnRails.setOptions({
  traceTurbolinks: TRACE_TURBOLINKS,
});
ReactOnRails.register({
  App,
});
ReactOnRails.registerStore({
  aStore,
});
`

`startup/App.jsx`:
`
import React, { PropTypes } from 'react';
import { Provider } from 'react-redux';
export default (_props, _railsContext) => {
  const aStore = ReactOnRails.getStore('aStore'); ///// this line causes the error
  return (
    <Provider store={ aStore }>
      <div></div>
    </Provider>
  );
};
`

`store/aStore.jsx`:
`
import { compose, createStore, applyMiddleware, combineReducers } from 'redux';
import thunkMiddleware from 'redux-thunk';
import reducers, { initialStates } from '../reducers';
export default(props, railsContext) => {
  const reducer = combineReducers(reducers);
  const composedStore = compose(
    applyMiddleware(thunkMiddleware)
  );
  const storeCreator = composedStore(createStore);
  const store = storeCreator(reducer, initialStates);
  return store;
};
`

`reducers/index.jsx`:
`
import bbbReducer, { initialState as bbbState } from './bbbReducer';
export default {
  bbbReducer,
};
export const initialStates = {
  bbbState,
};
`

`reducers/bbbReducer.jsx`:
`
export const initialState = [{ name: 'TEST', price: 123 }];
export default (state = initialState, action) => {
  switch(action.type) {
    case 'FETCH_ALL':
      return [
        {name: 'First',price: 100},{name: 'Second',price: 200},{name: 'Third',price: 300}
      ];
    default:
      return state;
  }
};
`


Error:
`
clientStartup.js?0ac5:108Uncaught Error: ReactOnRails encountered an error while rendering component: App.Original message: There are no stores hydrated and you are requesting the store commentsStore. This can happen if you are server rendering and you do not call redux_store near the top of your controller action's view (not the layout) and before any call to react_component.
`

**Comment by vsai** [comment]
I think I resolved the issue:

`ReactOnRails.registerStore` only is to be called when there is some server-side rendering involved. Because `registerStore` is to allow rails to access the store.
`ReactOnRails.stores()` returns only hydrated stores. And "hydration" only occurs on the server.

To "hydrate" client-side, you need to pass in the props through the App component, and use that to create the store: `const store = createStore(props);` and then pass that into the provider.

Can you confirm?

**Comment by justin808** [comment]
@vsai 
> ReactOnRails.registerStore only is to be called when there is some server-side rendering involved. 

Not true. Totally separate issues. `registerStore` is used if you want to use the shared redux store functionality. Server rendering (prerender: true) is a different thing.

**Comment by mnd-dsgn** [comment]
For me this issue came up because I forgot to put `redux_store_hydration_data` at the bottom of my main layout file. Try that.

**Comment by serodriguez68** [comment]
Hi!  I am just learning React on Rails and I want to make sure I am understanding the concepts right.

I'm running into the same error message described by @vsai on the original comment (No.2). If I use `registerStore` I *have* to use `redux_store` in the *view* regardless of server rendering to get the error message away.

1. Is `redux_store` necessary because of the store sharing functionality that `registerStore`gives?
2. If yes, why does `redux_store` only works in the *view* (not the controller) for shared store hydration purposes?

Thanks in advance for your help!

**Comment by justin808** [comment]
@serodriguez68 please give me the references in the docs that are not clear and I'll try to clarify.

Yes, you have to put redux_store (something like that) on the view so that we put the HTML element to populate the store.

**Comment by serodriguez68** [comment]
@justin808 thanks for your answer.  These are the part of the docs that I think can potentially cause confusion. 

Please take into account that I'm a total beginner with React On Rails (although I'm quite familiar with Rails); hence, I might just be misunderstanding the docs.

----------------------------

**1. Using `redux_store` in controller is better than in view VS Initializing Redux Store is optional VS `redux_store` must be used in the view for sharing a store with multiple components**
    1.1 https://github.com/shakacode/react_on_rails#controller-extension: States that the use of `redux_store`in the **CONTROLLER** is preferable to using the equivalent view helper.
    1.2 https://github.com/shakacode/react_on_rails#redux-store-notes: Suggests that using the `redux_store`helper is **OPTIONAL** (though recommended) if you want to share the store.  From my experience, this seems to be compulsory and only works if `redux_store` is used in the view.
    1.3 https://github.com/shakacode/react_on_rails#multiple-react-components-on-a-page-with-one-store: States that `redux_store`must be used in the **VIEW** for store sharing purposes.

**2. Multiple React Components on a Page with One Store: It is not clear why is it necessary to initialize `redux_store`  even without server rendering**.
    2.1 https://github.com/shakacode/react_on_rails#multiple-react-components-on-a-page-with-one-store
    2.2 I understand that initializing the store is desirable for being able to do a meaningful first-time client render of components.  However, one could theoretically get away with a 'Loading...' message on the containers if the store is empty and fetching info in `componentDidMount`.

Although the 'Loading...' option is probably ugly coding, people that como from a SPA background are used to it. Hence, I suggest clarifying why store initialization is **compulsory**  in react-on-rails.

-----------------------------------

Thanks for your help!  I'm deep diving into react-on-rails after diving deep into Rails 5.1 wepbacker install:react.  I find that react-on-rails is a much mature approach to using React and Redux on Rails. 

If you think my learning experience can be valuable for this project, let me know.  I would be more than happy to share my learning experience with you and your team in order to reduce the 'entry learning curve' for this project.
